clang/lib/Driver/Driver.cpp

```diff
@@ -4369,22 +4385,31 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
       LinkerInputs.push_back(TLI);
   }
   const llvm::opt::OptTable &Opts = getOpts();
-  if (C.getDefaultToolChain().getTriple().isWindowsMSVCEnvironment() &&
-      Args.hasArg(options::OPT_foffload_static_lib_EQ)) {
-    // In MSVC environment offload-static-libs are handled slightly different
-    // because of missing support for partial linking in the linker. We add an
-    // unbundling action for each static archive which produces list files with
-    // extracted objects. Device lists are then added to the appropriate device
-    // link actions and host list is ignored since we are adding
-    // offload-static-libs as normal libraries to the host link command.
+  auto unbundleStaticLib = [&](types::ID T) {
     for (const auto *A : Args.filtered(options::OPT_foffload_static_lib_EQ)) {
       Arg *InputArg = MakeInputArg(Args, Opts, A->getValue());
-      Action *Current = C.MakeAction<InputAction>(*InputArg, types::TY_Archive);
+      Action *Current = C.MakeAction<InputAction>(*InputArg, T);
       OffloadBuilder.addHostDependenceToDeviceActions(Current, InputArg, Args);
       OffloadBuilder.addDeviceDependencesToHostAction(
           Current, InputArg, phases::Link, PL.back(), PL);
     }
+  };
+  if (Args.hasArg(options::OPT_foffload_static_lib_EQ)) {
```

Comment:
For me, the following approach feels cleaner:
```suggestion
  for (const auto *A : Args.filtered(options::OPT_foffload_static_lib_EQ)) {
```
..., and then simply moving the lambda contents into the loop, setting a `types::ID` variable based on the same conditions and using this variable in the `MakeAction()` call

Reply:
There are cases where we will unbundle a particular lib more than once (e.g. Windows and -fintelfpga) so pulling in the unbundling wouldn't be possible without some code duplication.

Reply:
I see, thanks. However, the lambda could be alleviated from the `for` loop, and called within the `for` loop from my suggestion.

Reply:
Done.